chainindex/read.go

```diff
@@ -0,0 +1,170 @@
+package chainindex
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/types/ethtypes"
+)
+
+var (
+	headIndexedWaitTimeout = 5 * time.Second
+)
+
+func (si *SqliteIndexer) GetMaxNonRevertedTipset(ctx context.Context) (*types.TipSet, error) {
+	si.closeLk.RLock()
+	if si.closed {
+		return nil, ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	var tipsetKeyCidBytes []byte
+	err := si.getMaxNonRevertedTipsetStmt.QueryRowContext(ctx).Scan(&tipsetKeyCidBytes)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, nil
+		}
+		return nil, xerrors.Errorf("failed to get max non reverted tipset: %w", err)
+	}
+
+	tipsetKeyCid, err := cid.Cast(tipsetKeyCidBytes)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to cast tipset key cid: %w", err)
+	}
+
+	// Can this error out for reverted tipsets ?
+	return si.cs.GetTipSetByCid(ctx, tipsetKeyCid)
+}
+
+func (si *SqliteIndexer) GetCidFromHash(ctx context.Context, txHash ethtypes.EthHash) (cid.Cid, error) {
+	si.closeLk.RLock()
+	if si.closed {
+		return cid.Undef, ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	var msgCidBytes []byte
+
+	err := si.queryMsgCidFromEthHash(ctx, txHash, &msgCidBytes)
+	if err == sql.ErrNoRows {
+		err = si.waitTillHeadIndexedAndApply(ctx, func() error {
```

Comment:
```suggestion
		// not found, but may be in latest head, so wait for it and check
		err = si.waitTillHeadIndexedAndApply(ctx, func() error {
```

the logic here and the follow-up `err` check is a bit confusing, some inline docs would be nice

Reply:
Refactored this to make it easier to read.